trie/ctrie/ctrie.go

```diff
@@ -569,3 +688,173 @@ func bitCount(x uint32) uint32 {
+	if atomic.CompareAndSwapPointer(
+		(*unsafe.Pointer)(unsafe.Pointer(&in.main)),
+		unsafe.Pointer(old), unsafe.Pointer(n)) {
+		gcasComplete(in, n, ct)
+		return atomic.LoadPointer(prevPtr) == nil
+	}
+	return false
+}
+
+// gcasRead performs a GCAS-linearizable read of the I-node's main node.
+func gcasRead(in *iNode, ctrie *Ctrie) *mainNode {
+	m := (*mainNode)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&in.main))))
+	prev := (*mainNode)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&m.prev))))
+	if prev == nil {
+		return m
+	}
+	return gcasComplete(in, m, ctrie)
+}
+
+// gcasComplete commits the GCAS operation.
+func gcasComplete(i *iNode, m *mainNode, ctrie *Ctrie) *mainNode {
+	for {
+		if m == nil {
+			return nil
+		}
+		prev := (*mainNode)(atomic.LoadPointer(
+			(*unsafe.Pointer)(unsafe.Pointer(&m.prev))))
+		root := ctrie.rdcssReadRoot(true)
+		if prev == nil {
+			return m
+		}
+
+		if prev.failed != nil {
+			// Signals GCAS failure. Swap old value back into I-node.
+			fn := prev.failed
+			if atomic.CompareAndSwapPointer((*unsafe.Pointer)(unsafe.Pointer(&i.main)),
+				unsafe.Pointer(m), unsafe.Pointer(fn.prev)) {
+				return fn.prev
+			}
+			m = (*mainNode)(atomic.LoadPointer(
+				(*unsafe.Pointer)(unsafe.Pointer(&i.main))))
+			continue
+		}
+
+		if root.gen == i.gen && !ctrie.readOnly {
+			// Commit GCAS.
+			if atomic.CompareAndSwapPointer(
+				(*unsafe.Pointer)(unsafe.Pointer(&m.prev)), unsafe.Pointer(prev), nil) {
+				return m
+			}
+			continue
+		}
+
+		// Generations did not match. Store failed node on prev to signal
+		// I-node's main node must be set back to the previous value.
+		atomic.CompareAndSwapPointer(
+			(*unsafe.Pointer)(unsafe.Pointer(&m.prev)),
+			unsafe.Pointer(prev),
+			unsafe.Pointer(&mainNode{failed: prev}))
+		m = (*mainNode)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&i.main))))
+		return gcasComplete(i, m, ctrie)
+	}
+}
+
+// rdcssDescriptor is an intermediate struct which communicates the intent to
+// replace the value in an I-node and check that the root's generation has not
+// changed before committing to the new value.
+type rdcssDescriptor struct {
+	old       *iNode
+	expected  *mainNode
+	nv        *iNode
+	committed bool
+	token     int8
+}
+
+// HACK: see note in rdcssReadRoot.
+const rdcssToken int8 = -1
+
+// readRoot performs a linearizable read of the Ctrie root. This operation is
+// prioritized so that if another thread performs a GCAS on the root, a
+// deadlock does not occur.
+func (c *Ctrie) readRoot() *iNode {
+	return c.rdcssReadRoot(false)
+}
+
+// rdcssReadRoot performs a RDCSS-linearizable read of the Ctrie root with the
+// given priority.
+func (c *Ctrie) rdcssReadRoot(abort bool) *iNode {
+	r := atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&c.root)))
+	desc := (*rdcssDescriptor)(r)
+	// This is an awful and unsafe hack. Is there a better way to determine the
+	// type of a pointer?
```

Comment:
Would love to hear some ideas on how to do this better. Main issue is `unsafe.Pointer` does not retain any type information.